【OJT課題_自販機システム】システム概要, ユースケース図, イベントフローを作成しました

wnakayama opened the pull request:
ご確認をお願い致します。
●OJT課題の概要
　https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/master/README.md

●ユースケース図/ユースケース記述 (html版プレビュー)
http://htmlpreview.github.io/?https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/develop/UseCase/UseCase_Description.html

● ソースコード
　- ユースケース図　　UseCase/UseCase_Diagram.puml
　- ユースケース記述　UseCase/UseCase_Description.adoc
　https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/1/files

wnakayama committed:
create diagram

wnakayama committed:
export to PNG

wnakayama committed:
create description

wnakayama committed:
add image block

iwanam reviewed (changes requested):
* 今のユースケースだと、飲み物とチケットを同時に販売するシステムになってしまいます
  - ヒント：基本的な自動販売機システムをベースにまず考えてみるといいかも？
* とりあえず、ユースケース記述には、イベントフローとかそのあたりを書いてくださいませ

SeigoKawamura commented on UseCase/UseCase_Diagram.puml:
```diff
@@ -0,0 +1,31 @@
+@startuml
+
+title UseCase_Diagram
```
もしタイトルを書くならば「1．ユースケース図」のUseCase_Diagramだと情報がないです。

SeigoKawamura commented on UseCase/UseCase_Diagram.puml:
```diff
@@ -0,0 +1,31 @@
+title UseCase_Diagram
+:ユーザー: as user
+
+rectangle システム {
```
```suggestion
rectangle 自販機システム {
```

SeigoKawamura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,60 @@
+:revnumber: 0.1
+:revdate: 2019/10/11
+
+= 自販機システム-機能要件
```
自販機システムとはどのようなシステムで何ができるようになるのですか？
いきなりユースケース図が出てきてよく分からかったです。

SeigoKawamura commented on UseCase/UseCase_Diagram.puml:
```diff
@@ -0,0 +1,31 @@
+    UC02
+    end note
+
+    note right of sellTicket #Aquamarine
```
優先度によって色を変えていると思うのですが、何も説明がないので分からないです。

tmatsumura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,60 @@
+|高
+
+|UC03
+|チケットを販売する
```
チケットとは何か（ユーザーはチケットを受け取ると何ができるのか、など）仕様書には説明がないです

tmatsumura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,60 @@
+
+|UC03
+|チケットを販売する
+|投入金額に応じて以下の2点を記録したチケットを提供する
```
何を購入できる（受け取れる）のかの情報は不要ですか？

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,60 @@
+|**優先度**
+
+|UC01
+|飲み物を販売する
```
ユースケース図は、 `アクター` が `use_case` する、と読めるように書きます。
この場合、
ユーザーは、飲み物を販売するとなり、おかしくないですか?(ユーザーは販売しないです。。)

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,60 @@
+
+|UC01
+|飲み物を販売する
+|投入金額に応じて飲み物を提供する
```
投入金額に応じて、出てくる飲み物が変わるんでしょうか?

tmatsumura replied:
https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/1#pullrequestreview-300469379 でも指摘されていますが、投入金額が足りてない場合はどうなるのかなど、いろんなケースがあると思うので、どういうフローで飲み物を提供するのか考えてみた方がいいです

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,60 @@
+|チケットを販売する
+|投入金額に応じて以下の2点を記録したチケットを提供する
+
+* 購入年月日
```
チケットの名称は必須です

skashiyama reviewed (changes requested):
PRのタイトルは、内容がわかるようにしてください。(初版を作成しました、だと何のPRかわからないです)

指摘についてですが、こう直したらいいとかはあえて書いてないです。どう修正するか考えてみてください。

wnakayama committed:
rename title and usecase

wnakayama committed:
export to PNG (rename title and usecase)

wnakayama committed:
add explanation and eventflow

wnakayama committed:
add explanation and eventflow

W-Nakayama-24 commented:
11日に頂いたレビューを踏まえて、以下の作業を実施しました。
-  機能ごとのイベントフローを記載
- 「ユースケース図」の前に「システム概要」を追加
- 「アクター 」が 「use_case 」する、と読めるように各ユースケースの記述を修正
- プルリクエストのタイトルを修正
-  機能に関する情報の追加・補足
- 自販機システムで購入できるものを「飲み物」に限定 (※)

※
課題ではWant要件として「自動券売機に応用できる設計」を目指していますが、
今回は大前提となる「飲み物を購入できるシステムの設計」を優先しました。
ご指摘いただいたチケットの名称、説明については後ほど検討させていただきます。

ご確認をお願い致します。

- ユースケース図/ユースケース記述 (html版プレビュー)
http://htmlpreview.github.io/?https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/develop/UseCase/UseCase_Description.html

- ソースコード
    - ユースケース図　　UseCase/UseCase_Diagram.puml
　- ユースケース記述　UseCase/UseCase_Description.adoc
　https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/1/files

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,314 @@
+システムメッセージ02
+....
+---[2]商品を購入する---
+商品名を入力してください
```
名前はユーザービリティ的に面倒なので、商品名に対応する数字を入力してもらえればいいかと思います。

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,314 @@
+
+=== 商品を購入する
+
+[cols="1a,3a"]
```
見やすさ的な意味で、ヘッダーの色は変えてほしいです

tmatsumura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,314 @@
+|
+. 利用者は自販機システムの実行ファイルを指定してプログラムを実行する
+. 初期画面が表示される。
+. 利用者は1,2,3のいずれかを入力する。
```
1,2,3以外が入力されたらどうなりますか？

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,314 @@
+. 利用者は自販機システムの実行ファイルを指定してプログラムを実行する
+. 初期画面が表示される。
+. 利用者は1,2,3のいずれかを入力する。
+. 自販機システムは入力された数字に対応したユースケースの機能を始動する
```
不要です。
初期画面の表示なので、初期画面が表示されればシナリオとしてokです。
というより、line 74 で終わらせないと事後条件とマッチしないです

tmatsumura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,314 @@
+|
+. 利用者はコンソール上で「2」を入力する
+. 自販機システムはシステムメッセージ02を表示し、文字列入力を受け付ける
+. 利用者は商品名を入力する
```
どの時点で代替フローに移動するか書いておいた方がわかりやすいと思います（このあたりからですかね？）

tmatsumura reviewed (changes requested):
全ての商品の在庫が0本になったらどうなりますか？

SeigoKawamura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,314 @@
+|初期画面が表示されている
+
+|**補足**
+|初期画面のイメージは以下の通りである
```
初期画面のイメージは別のセクションに移動して`初期画面`にそのセクションのリンクを貼ると見やすいかもしれませんね。機能詳細に`初期画面`が頻繁に出てくるので。

iwanam commented:
ちなみに全員の確認が終わらなくても、直すべきと思ったとこはどんどん直していっていいです。良くない状態のまま他の人のレビューを待っても、結局大きく直したら意味ないですからね。
他の人の意見も聞いてから考えたい、みたいなとこは待っててもいいですけど

wnakayama committed:
ADD Product Num, Alt Flow, Header Color etc.

ユーザビリティ向上のため、商品番号・ヘッダーカラー・[初期画面]セクションへの内部参照を追加
 「全ての商品が在庫切れである時」の代替フローを追加
代替フローへ移行する条件を基本フローの行に追加

W-Nakayama-24 commented:
以下の作業を実施しました。

- 商品の指定を「商品名」で受け付ける方式から、「商品番号」で受け付ける方式へ変更
- テーブルのヘッダー列の色を変更
- 「システム始動」-初期画面の表示について、基本フローのシナリオを修正

- 「すべての商品の在庫が0本になった場合」の代替フローをUC02に追加

- 「どの時点で代替フローに移動するか」の説明を各ユースケースの基本フローに追加

- 「初期画面」の補足説明を別個のセクションに移動、各ユースケースの説明から「初期画面」セクションに移動する内部参照のリンクを追加

※
**システム始動後、各機能を呼び出す際に「1,2,3以外が入力された場合」の対応について**
実装時にはエラーメッセージを出すことで対応したいと思っていますが、この説明をどの箇所に盛り込めばよいか判断しきれませんでした。
何かヒントやご提案がありましたら、お聞かせいただけると幸いです。

よろしくお願い致します。


- ユースケース図/ユースケース記述 (html版プレビュー)

http://htmlpreview.github.io/?https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/develop/UseCase/UseCase_Description.html

- ソースコード

    - ユースケース図　　UseCase/UseCase_Diagram.puml

    - ユースケース記述　UseCase/UseCase_Description.adoc

　https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/1/files

tmatsumura commented:
> ※
**システム始動後、各機能を呼び出す際に「1,2,3以外が入力された場合」の対応について**
実装時にはエラーメッセージを出すことで対応したいと思っていますが、この説明をどの箇所に盛り込めばよいか判断しきれませんでした。

3.2～4のユースケースは1,2,3が入力された後のフローなので、それより前の3.1.1の初期画面に説明を足すのがいいかなと思います

SeigoKawamura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,333 @@
+|
+[[alt_uc01]]
+. コンソールの入力内容が 1000, 500, 100, 50, 10 以外の数値だった場合
+** エラーメッセージ01をコンソールに表示する 
```
他のケースもそうですが、代替フローに移行してエラーメッセージをコンソールに表示した後の処理はどうなりますか？

wnakayama replied:
「代替フローは実行しても事後条件を満たす」
「例外フローは実行すると事後条件が満たされない」
という認識のもとで、いずれの代替フローも実行した後は初期画面へ戻る仕様にするつもりです。

skashiyama reviewed (changes requested):
以下、質問です。

- プログラムはどうやったら終了しますか?(その際の投入金額はどうなりますか?)
- プログラムを終了した後、再度実行した場合、在庫はどうなりますか?(引き継がれる? or 初期状態に戻る?)

どう対応するか、対応した場合にどれくらい時間がかかりそうかなどもあるので相談しながらやるといいかもです。

W-Nakayama-24 commented:
> * プログラムはどうやったら終了しますか?(その際の投入金額はどうなりますか?)

考慮できていませんでした… ご指摘ありがとうございます。
現段階では、初期画面に **"[9] システムを終了する"** を追加して
コンソールから数字を入力させることで終了できるようにしようと思います。

プログラム終了時には、投入金額も0円にリセットする形を取ろうと思います。


> * プログラムを終了した後、再度実行した場合、在庫はどうなりますか?(引き継がれる? or 初期状態に戻る?)

初期状態へ戻る仕様にしようと考えています。


> どう対応するか、対応した場合にどれくらい時間がかかりそうかなどもあるので相談しながらやるといいかもです。

アドバイスありがとうございます。こまめに先輩方の知恵をお借りしたいと思います。

wnakayama committed:
ADD Alt Flow, End of Program, End of AltFlow

・機能呼び出しで「1,2,3,9」以外を入力したときの代替フローを追加
・機能「システム終了」を追加
・エラーメッセージ出力後の処理について説明を追加

wnakayama committed:
ADD internal reference "End of Program"

 line:126 「システム終了」に内部相互参照のリンクを追加

W-Nakayama-24 commented:
以下の作業を実施しました。

・機能呼び出しで「1,2,3,9」以外を入力したときの代替フローをセクション3.1.1「初期画面」に追加
・機能「システム終了」を追加
・プログラム終了/再実行時の「投入金額」「商品の在庫」の処理について説明を追加
・エラーメッセージ出力後の、基本フローへ戻る処理について説明を追加

よろしくお願い致します。

- ユースケース図/ユースケース記述 (html版プレビュー)
http://htmlpreview.github.io/?https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/develop/UseCase/UseCase_Description.html

- ソースコード(差分)
 https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/1/commits/888f6f92c65b378951a485ad65db1000ce6a6658
https://github.com/W-Nakayama-24/VendingMachine_CUI/pull/1/commits/0120c6c97ba6b9ba2a200e7c54bb63b5ad030711
　

skashiyama commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,424 @@
+
+|補足
+|
+* 自販機システム上に残った「投入金額」の数値は、0にリセットされる。
```
投入した金額が戻ってこないってことになりません?
金額が0でないと9を受け付けないようにした方が親切だと思います。

tmatsumura commented on UseCase/UseCase_Description.adoc:
```diff
@@ -0,0 +1,424 @@
+[[alt_uc01]]
+. コンソールの入力内容が 1000, 500, 100, 50, 10 以外の数値だった場合
+.. エラーメッセージ01をコンソールに表示する
+.. 基本フローのステップ6に移動する(シナリオを終了する)
```
- どちらでもいいですが、ステップ2に戻った方が使いやすいような気がしました
- 他の代替フローもそういうケースがあるかもしれないので確認してみた方がいいと思います

tmatsumura replied:
投入のキャンセルとか考えないといけない気もするのでこのままでもいい気がしました

SeigoKawamura reviewed (approved):
私はOKです。

wnakayama committed:
ADD exFlow_systemEnd. FIX altFlow_uc01

・「システム終了」 投入金額が残っている場合の例外フローを追加
・「UC01 お金を投入する」 入力が正しくない場合の代替フローを修正

W-Nakayama-24 commented:
以下の作業を実施しました。

・**「システム終了」機能**
　　投入金額が残っている場合において、プログラムを終了せずに
　　返金の受け取りを促す例外フローを追加
　
・**「UC01 お金を投入する」機能** 
　　文字列、空文字列の入力にも対応できるように代替フロー1の移動条件を修正
　　数字の入力を再度受け付けるように代替フロー1の内容を修正
　　
・**「UC02 商品を購入する」機能**
　　文字列の入力にも対応できるように代替フロー2の移動条件を修正 

・「数字」「数値」、「価格」「単価」など表記ゆれがあった部分を統一

よろしくお願い致します。

- ユースケース図/ユースケース記述 (html版プレビュー)
http://htmlpreview.github.io/?https://github.com/W-Nakayama-24/VendingMachine_CUI/blob/develop/UseCase/UseCase_Description.html

- ソースコード
e05f6077628fca428f7a8fc0fea75864ed56f228

W-Nakayama-24 commented:
マージしました。ありがとうございました!
